plugins/dev9ghzdrk/pcap_io.cpp

```diff
@@ -75,53 +80,76 @@ int GetMACAddress(char *adapter, mac_address* addr)
 		pAdapterInfo = pAdapterInfo->Next;    // Progress through
 	}
 	while(pAdapterInfo);                    // Terminate if last adapter
-	return 0;
+#elif defined(__linux__)
+	struct ifreq ifr;
+	int fd = socket(AF_INET, SOCK_DGRAM, 0);
+	strcpy(ifr.ifr_name, adapter);
+	if (0 == ioctl(fd, SIOCGIFHWADDR, &ifr))
+	{
+		retval = 1;
+		memcpy(addr,ifr.ifr_hwaddr.sa_data,6);
+	}
+	else
+	{
+		SysMessage("Could not get MAC address for adapter: %s", adapter);
+	}
+	close(fd);
+#endif
+	return retval;
 }
 
 int pcap_io_init(char *adapter)
 {
+	struct bpf_program fp;
+	char filter[1024] = "ether broadcast or ether dst ";
 	int dlt;
 	char *dlt_name;
 	emu_printf("Opening adapter '%s'...",adapter);
 	u16 checksum;
 	GetMACAddress(adapter,&host_mac);
 	
-	//Near copy of the host mac, butchered slightly, should be pretty good!
-	eeprom[0] = host_mac.bytes[0];
-	eeprom[1] = host_mac.bytes[1];
-	eeprom[2] = host_mac.bytes[2];
-	eeprom[3] = host_mac.bytes[2];
-	eeprom[4] = host_mac.bytes[5];
-	eeprom[5] = host_mac.bytes[4];
-
-	virtual_mac.bytes[0] = host_mac.bytes[0];
-	virtual_mac.bytes[1] = host_mac.bytes[1];
-	virtual_mac.bytes[2] = host_mac.bytes[2];
-	virtual_mac.bytes[3] = host_mac.bytes[2];
-	virtual_mac.bytes[4] = host_mac.bytes[5];
-	virtual_mac.bytes[5] = host_mac.bytes[4];
+	//Lets take the hosts last 2 bytes to make it unique on Xlink
+	virtual_mac.bytes[4] = host_mac.bytes[4];
+	virtual_mac.bytes[5] = host_mac.bytes[5];
+
+	for(int ii=0; ii<6; ii++)
+		eeprom[ii] = virtual_mac.bytes[ii];
```

Comment:
@TheLastRar could you please test ability to connect to server again? I've updated the bit to copy MAC into the BIOS so that there's no need for address translation during each send and receive (send and receive have been simplified accordingly)